rfcs/192-fr-copy-receipt-performance.md

```diff
@@ -0,0 +1,177 @@
+}
+```
+
+The `GetCountOfCopiesAsync` method of the SQLite `JournalFRCopyReceiptsRepository` looks like this:
+```cs
+public async Task<int> GetCountOfCopiesAsync(string cbPreviousReceiptReference) {
+    var query = "SELECT count(*) AS ReceiptCase FROM ftJournalFRCopyReceipts WHERE cbPreviousReceiptReference = @cbPreviousReceiptReference";
+    return await DbConnection.QuerySingleAsync<int>(query, new { cbPreviousReceiptReference });
+}
+```
+
+## Migration
+
+The new table will be created with the usual migration process.
+
+On start up of the Queue (in the 1.2 `workerFR.cs`) we check if the table contains any entries and if not we will populate it with the old data from the `ftJournalFR` table.
+
+The population of the table will be done like this:
+
+```cs
+private async Task<int> PopulateFtJournalFRCopyReceiptsTableAsync()
+{
+    await foreach (var copyJournal in _journalFRRepository.GetProcessedCopyReceiptsAsync())
+    {
+        var queueItem = await _queueItemRepository.GetAsync(copyJournal.ftQueueItemId);
+        var request = queueItem?.request != null ? JsonConvert.DeserializeObject<ReceiptRequest>(queueItem.request) : null;
+        var response = queueItem?.response != null ? JsonConvert.DeserializeObject<ReceiptResponse>(queueItem.response) : null;
+        if (response != null)
+        {
+            var duplicata = response.ftSignatures.FirstOrDefault(x => x.Caption == "Duplicata");
+
+            if (duplicata != null)
+            {
+                _journalFRCopyReceiptsRepository.Insert(new FtJournalFRCopyReceipt
+                {
+                    cbPreviousReceiptReference = request.cbPreviousReceiptReference,
+                    ftQueueItemId = copyJournal.ftQueueItemId
+                });
+            }
+        }
+    }
+
+    return count;
+}
+```
+
+## SignatureCloud
+
+We will setup a new SignatureCloud environment which uses the new implementation and keep using the old one in the beginning.
+We will have a load balancer in front of the SignatureCloud which will route the requests to the correct environment.
+
+When the PosCreator is ready to use the new implementation we will switch the load balancer to the new environment.
+This will be done in coordination with the PosCreators and the recertification.
+
+### AzureTableStorage
+
+The Azure Table Storage Queue will have the `cbPreviousReceiptReference` as a Partition key and the `ftQueueItemID` as a Row key.
+
+This means counting all rows with the same `cbPreviousReceiptReference` will be very fast.
+
+# Drawbacks
+
+This solution involves a new table and we need to populate it with the data from the `ftJournalFR` table before we can use the new implementation.
+This brings some problems with it especially in the SignatureCloud.
+
+This update version of the middleware might need to be certified again.
+
+# Rationale and alternatives
+
+An alternative way of doing this would be to keep the current implementation and improve it.
```

Comment:
@TSchmiedlechner @steininger even though this option wouldn't necessarily solve all of our problems it would be very little effort and could already improve performance hugely. Curious about your opinion

Reply:
I think if it's possible it would be a nice intermediate solution that would relive some pressure and give us more time to deal with the situation.

Reply:
Interesting idea, why not :) I'd make sure to not "certify" the intermediate solution though, so we'd maybe have multiple RCs.

Reply:
I've just thought some more about that and it would not really help (in this form) since the first copyreceipt of a receipt would always iterate through all receipts and then return 0... we would also have to check for normal receipts and check their receiptreferences and stop with count 0 if we find one. That would maybe change the behaviour of the implementation in certain edge cases.

Reply:
As far as certification goes,  if i understood correctly we will use self-testation for the RCs

Reply:
> Interesting idea, why not :) I'd make sure to not "certify" the intermediate solution though, so we'd maybe have multiple RCs.

So as far as I understood @Bluette-C-Riviere we will anyways have to choose the self testation part regardless of what we do. So to show some progress we could implement this suggested solution in rc1 roll this out for sqlite and see if it solves the most urgent issues. We then could keep on looking for the approach described elarier that involves a lookup table